sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,436 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Protocol Buffers describing the Runner API, which is the runner-independent,
+ * SDK-independent definition of the Beam model.
+ */
+
+syntax = "proto3";
+
+package org.apache.beam.runner_api.v1;
+
+option java_package = "org.apache.beam.sdks.common.runner_api.v1";
+option java_outer_classname = "RunnerApi";
+
+import "google/protobuf/any.proto";
+import "google/protobuf/timestamp.proto";
+
+// A Pipeline contains the entire hierarchical graph of transforms as well
+// as descriptors for all PCollections.
+//
+// The hierarchical graph itself is an independent layer from the payloads
+// specific to the Runner API, for ease of reuse by e.g. the Fn API.
+message Pipeline {
+
+  // (Required) A map from pipeline-scoped id to graph node.
+  //
+  // Each node is required to contain a PTransform specification.
+  map<int64, GraphNode> transform_nodes = 1;
+
+  // (Required) A map from pipeline-scoped id to PCollection.
+  map<int64, PCollection> pcollections = 2;
+
+  // (Required) A map from pipeline-scoped id to WindowingStrategy.
+  map<int64, WindowingStrategy> windowing_strategies = 3;
+
+  // (Required) A map from pipeline-scoped id to Coder.
+  map<int64, Coder> coders = 4;
+
+  // (Required) A map from pipeline-scoped id to Environment.
+  map<int64, Environment> environments = 5;
+
+  // (Optional) Map from option name to arbitrary serialization of its value.
+  //
+  // Some options may be user-specific or SDK-specific, in which case the
+  // payload is expected to be unintelligible.
+  //
+  // Some options may be standardized across SDKs, in which case the universe
+  // of possible values is determined by the name of the pipeline option.
+  //
+  // TODO: document what is standardized
+  map<string, google.protobuf.Any> pipeline_options = 6;
+
+  // (Optional) Static display data for the pipeline.
+  map<string, string> display_data = 7;
+
+  // TODO: consider identifying the SDK + runner shim submitting the pipeline
+}
+
+// A generic node in a bipartite directed hierarchical graph.
+message GraphNode {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A URN that describes what kind graph node this is.
+  //
+  // Specifically, the URN should be enough to decipher the inputs,
+  // outputs, and payload.
+  string urn = 2;
```

Comment:
We should discuss whether we really want to use an URN here. With proto, we have the ability to do something like:

message GenericKind {
  string urn = 1;
  ... generic arguments ...
}

message CoGroupByKey { ... }
etc.

oneof NodeKindAndPayload {
  GenericKind generic_kind = 2;
  CoGroupByKey cogbk_kind = 3;
  ...
}

etc.

Do we care about what the generated JSON actually looks like? If so, that proto structure allows more descriptive field names and such in the generated JSON.